core/src/main/scala/com/salesforce/op/stages/impl/feature/HumanNameDetector.scala

```diff
@@ -0,0 +1,162 @@
+
+    new HumanNameDetectorModel[T](uid, treatAsName, orderedGenderDetectStrategies)
+  }
+}
+
+class HumanNameDetectorModel[T <: Text]
+(
+  override val uid: String,
+  val treatAsName: Boolean,
+  val orderedGenderDetectStrategies: Seq[GenderDetectStrategy] = Seq.empty[GenderDetectStrategy]
+)(implicit tti: TypeTag[T])
+  extends UnaryModel[T, NameStats]("human name detector", uid) with NameDetectFun {
+
+  import NameStats.BooleanStrings._
+  import NameStats.GenderStrings.GenderNA
+  import NameStats.Keys._
+  def transformFn: T => NameStats = (input: T) => {
+    val tokens = preProcess(input.value)
+    if (treatAsName) {
+      require(orderedGenderDetectStrategies.nonEmpty, "There must be a gender extraction strategy if treating as name.")
+      // TODO: Figure out how to use a broadcast variable for the gender dictionary within a unary transformer
+      val genders: Seq[String] = orderedGenderDetectStrategies map {
+        identifyGender(input.value.getOrElse(""), tokens, _, NameDetectUtils.DefaultGenderDictionary)
+      }
+      val gender = genders.find(_ != GenderNA).getOrElse(GenderNA)
+      NameStats(Map(
+        IsNameIndicator -> True,
+        OriginalName -> input.value.getOrElse(""),
+        Gender -> gender
+      ))
+    }
+    else NameStats(Map.empty[String, String])
+  }
+}
+
+case class HumanNameDetectorMetadata
+(
+  treatAsName: Boolean,
+  predictedNameProb: Double,
+  genderResultsByStrategy: Map[String, GenderStats]
+) extends MetadataLike {
+  import HumanNameDetectorMetadata._
+
+  override def toMetadata(): Metadata = {
+    val metaDataBuilder = new MetadataBuilder()
+    metaDataBuilder.putBoolean(TreatAsNameKey, treatAsName)
+    metaDataBuilder.putDouble(PredictedNameProbKey, predictedNameProb)
+    val genderResultsMetaDataBuilder = new MetadataBuilder()
+    genderResultsByStrategy foreach { case (strategyString, stats) =>
+      genderResultsMetaDataBuilder.putDoubleArray(strategyString, Array(stats.numMale, stats.numFemale, stats.numOther))
+    }
+    metaDataBuilder.putMetadata(GenderResultsByStrategyKey, genderResultsMetaDataBuilder.build())
+    metaDataBuilder.build()
+  }
+
+  override def toMetadata(skipUnsupported: Boolean): Metadata = toMetadata()
+}
+
+case object HumanNameDetectorMetadata {
+  val TreatAsNameKey = "treatAsName"
+  val PredictedNameProbKey = "predictedNameProb"
+  val GenderResultsByStrategyKey = "genderResultsByStrategy"
+
+  def fromMetadata(metadata: Metadata): HumanNameDetectorMetadata = {
+    val genderResultsMetadata = metadata.getMetadata(GenderResultsByStrategyKey)
+    val genderResultsByStrategy: Map[String, GenderStats] = {
+      NameDetectUtils.GenderDetectStrategies map { strategy: GenderDetectStrategy =>
+        val strategyString = strategy.toString
+        val resultsArray = genderResultsMetadata.getDoubleArray(strategyString)
+        require(resultsArray.length == 3, "There must be exactly three values for each gender detection strategy.")
```

Comment:
minor: name the values - `s"There must be exactly three values for each gender detection strategy: $TreatAsNameKey, $PredictedNameProbKey, $GenderResultsByStrategyKey"`